hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+//! // Automatic Switch to Compare is disabled
+//! icm.set_ascd(false);
+//!
+//! // Setup region 0 to monitor memory
+//! icm_region_desc
+//!     .region0
+//!     .set_raddr(&message_region0_sha1);
+//! icm_region_desc.region0.rcfg.reset_rcfg_to_default();
+//! icm_region_desc.region0.rcfg.set_algo(icm_algorithm::SHA1);
+//! // Activate Compare Digest (should be true when comparing memory)
+//! icm_region_desc.region0.rcfg.set_cdwbn(true);
+//! // Digest Mismatch Interrupt Disable (enabled)
+//! icm_region_desc.region0.rcfg.set_dmien(false);
+//!
+//! // Set Region Mismatch Interrupt
+//! icm_region0.set_rdm_int();
+//!
+//! // Setup region 1 to monitor memory
+//! icm_region_desc
+//!     .region1
+//!     .set_raddr(&message_region1_sha1);
+//! icm_region_desc.region1.rcfg.reset_rcfg_to_default();
+//! icm_region_desc.region1.rcfg.set_algo(icm_algorithm::SHA1);
+//! // Activate Compare Digest (should be true when comparing memory)
+//! icm_region_desc.region1.rcfg.set_cdwbn(true);
+//! // Digest Mismatch Interrupt Disable (enabled)
+//! icm_region_desc.region1.rcfg.set_dmien(false);
+//!
+//! // Set Region Mismatch Interrupt
+//! icm_region1.set_rdm_int();
+//!
+//! // Setup region 2 to monitor memory
+//! icm_region_desc
+//!     .region2
+//!     .set_raddr(&message_region2_sha224);
+//! icm_region_desc.region2.rcfg.reset_rcfg_to_default();
+//! icm_region_desc.region2.rcfg.set_algo(icm_algorithm::SHA224);
+//! // Activate Compare Digest (should be true when comparing memory)
+//! icm_region_desc.region2.rcfg.set_cdwbn(true);
+//! // Digest Mismatch Interrupt Disable (enabled)
+//! icm_region_desc.region2.rcfg.set_dmien(false);
+//!
+//! // Set Region Mismatch Interrupt
+//! icm_region2.set_rdm_int();
+//!
+//! // Setup region 3 to monitor memory
+//! icm_region_desc
+//!     .region3
+//!     .set_raddr(&message_region3_sha256);
+//! icm_region_desc.region3.rcfg.reset_rcfg_to_default();
+//! icm_region_desc.region3.rcfg.set_algo(icm_algorithm::SHA256);
+//! // Activate Compare Digest (should be true when comparing memory)
+//! icm_region_desc.region3.rcfg.set_cdwbn(true);
+//! // Digest Mismatch Interrupt Disable (enabled)
+//! icm_region_desc.region3.rcfg.set_dmien(false);
+//! // Wrap
+//! icm_region_desc.region3.rcfg.set_wrap(true);
+//!
+//! // Set Region Mismatch Interrupt
+//! icm_region3.set_rdm_int();
+//!
+//! // Modify regions to trigger interrupts
+//! message_region0_sha1[3] = 0xDEAD_BEEF;
+//! message_region1_sha1[4] = 0xDEAD_BEEF;
+//! message_region2_sha224[5] = 0xDEAD_BEEF;
+//! message_region3_sha256[6] = 0xDEAD_BEEF;
+//!
+//! icm.enable()
+use crate::pac::generic::Variant;
+use crate::pac::icm::uasr::URAT_A;
+
+use crate::pac::icm::*;
+use crate::typelevel::Sealed;
+use core::marker::PhantomData;
+
+/// Reexport the User SHA Algorithm
+pub use crate::icm::cfg::UALGO_A as icm_algorithm;
+
+/// IcmToken represents the peripheral and it encapsulates the hardware register
+struct IcmToken {
+    icm: crate::pac::ICM,
+}
```

Comment:
What exactly is the purpose of this struct? It's similar to the `Registers` structs in `v2::spi` and `v2::uart`, but there's a key difference. The `Registers` structs act to *completely* encapsulate interaction with the PAC, which means calls to PAC functions are contained to a single type. But this struct only forwards along references to the particular registers.

I also don't know that `Token` is the right name. Intuitively, I would expect a `Token` type to have no real utility on its own. It can only be exchanged for something that does have utility. I always imagine it like the tokens at a video game arcade. We use a lot of `Token` types in the `clock::v2` module, but there it makes more sense. The PAC does not have any dedicated types for the individual clocks, so we create various `Token`s to represent uninitialized clocks.

In this case, I think the `pac::ICM` type already acts as the token. Users could use the PAC struct directly, or they can exchange it for a `hal::Icm` struct.

If it were me, I would either: 1) remove this type and access the `pac::ICM` struct from the `hal::Icm` type directly, or 2) turn this into a dedicated register interface struct, and remove all PAC register writing from other types.

Reply:
I agree it was somewhere in-between, not ideal. Opted for alternative 1), since `Region` also needs access to `ICM` and it isn't easy (at least to my knowledge) to refer back to this single copy it would be a bit clunky (imho) passing around this dedicated register interface.

tl;dr: `pac::ICM` is the "token", no need for extra "token"